bothan-core/src/manager/crypto_asset_info/price/tasks.rs

```diff
@@ -0,0 +1,180 @@
+use num_traits::Zero;
+use rust_decimal::Decimal;
+use std::collections::{HashSet, VecDeque};
+use thiserror::Error;
+use tracing::{error, info, warn};
+
+use crate::manager::crypto_asset_info::price::cache::PriceCache;
+use crate::manager::crypto_asset_info::types::{PriceState, WorkerMap};
+use crate::registry::post_processor::{PostProcess, PostProcessError};
+use crate::registry::processor::{Process, ProcessError};
+use crate::registry::signal::Signal;
+use crate::registry::source::OperationRoute;
+use crate::registry::{Registry, Valid};
+use crate::store::ActiveSignalIDs;
+use crate::worker::AssetState;
+
+#[derive(Debug, Error)]
+enum Error {
+    #[error("Signal does not exist")]
+    InvalidSignal,
+
+    #[error("Prerequisites required: {0:?}")]
+    PrerequisiteRequired(Vec<String>),
+
+    #[error("Failed to process signal: {0}")]
+    FailedToProcessSignal(#[from] ProcessError),
+
+    #[error("Failed to post process signal: {0}")]
+    FailedToProcessPostSignal(#[from] PostProcessError),
+}
+
+pub async fn get_signal_price_states<'a>(
+    ids: Vec<String>,
+    workers: &WorkerMap<'a>,
+    registry: &Registry<Valid>,
+    active_signal_ids: &ActiveSignalIDs,
+    stale_cutoff: i64,
+) -> Vec<PriceState> {
+    let mut cache = PriceCache::new();
+
+    let mut queue = VecDeque::from(ids.clone());
+    while let Some(id) = queue.pop_front() {
+        if cache.contains(&id) {
+            continue;
+        }
+
+        match compute_signal_result(&id, workers, registry, stale_cutoff, &cache).await {
+            Ok(price) => {
+                info!("signal {}: {} ", id, price);
+                cache.set_available(id, price);
+            }
+            Err(Error::PrerequisiteRequired(prereqs)) => {
+                info!("prerequisites required for signal {}: {:?}", id, prereqs);
+                for prereq in prereqs {
+                    queue.push_front(prereq)
+                }
+                queue.push_back(id);
+            }
```

Comment:
I think it should be something like this. otherwise, it might loop more than it should for 2-level dependency. (if I'm not wrong...)

for example, A needs B, B needs C. This dependency will have the result as below.
step | queue
1| [A]
2| [B, A]
3| [C, A, B]
4| [A, B]
5| [B, B, A]



```suggestion
            Err(Error::PrerequisiteRequired(prereqs)) => {
                info!("prerequisites required for signal {}: {:?}", id, prereqs);
                queue.push_front(id);
                for prereq in prereqs {
                    queue.push_front(prereq)
                }
            }
```